server/src/test/java/io/spine/server/storage/EntityRecordStorageTest.java

```diff
@@ -0,0 +1,603 @@
+                           .build();
+    }
+
+    @Test
+    @DisplayName("retrieve empty iterator if storage is empty")
+    void retrieveEmptyIterator() {
+        FieldMask nonEmptyFieldMask = FieldMask
+                .newBuilder()
+                .addPaths("invalid-path")
+                .build();
+        EntityRecordStorage<StgProjectId, ?> storage = storage();
+        RecordQuery<StgProjectId, EntityRecord> query =
+                storage.queryBuilder()
+                       .withMask(nonEmptyFieldMask)
+                       .build();
+        Iterator<?> empty = storage.readAll(query);
+
+        assertNotNull(empty);
+        assertFalse(empty.hasNext(), "Iterator is not empty!");
+    }
+
+    @Test
+    @DisplayName("delete record")
+    void deleteRecord() {
+        EntityRecordStorage<StgProjectId, StgProject> storage = storage();
+        StgProjectId id = newId();
+        EntityRecord record = newStorageRecord(id);
+
+        // Write the record.
+        storage.write(id, record);
+
+        // Delete the record.
+        assertTrue(storage.delete(id));
+
+        // There's no record with such ID.
+        assertFalse(storage.read(id)
+                           .isPresent());
+    }
+
+    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection" /* Storing of generated objects and
+                                                                checking via #contains(Object). */)
+    @Test
+    @DisplayName("create unique states for same ID")
+    void createUniqueStatesForSameId() {
+        int checkCount = 10;
+        StgProjectId id = newId();
+        Set<EntityState<StgProjectId>> states = newHashSet();
+        for (int i = 0; i < checkCount; i++) {
+            EntityState<StgProjectId> newState = newState(id);
+            if (states.contains(newState)) {
+                fail("RecordStorageTest.newState() should return unique messages.");
+            }
+        }
+    }
+
+    @Nested
+    @DisplayName("read")
+    class Read {
+
+        @Test
+        @DisplayName("single record according to the specific field mask")
+        void singleRecord() {
+            StgProjectId id = newId();
+            EntityRecord record = newStorageRecord(id);
+            EntityRecordStorage<StgProjectId, StgProject> storage = storage();
+            storage.write(id, record);
+
+            EntityState<StgProjectId> state = newState(id);
+            FieldMask idMask = fromFieldNumbers(state.getClass(), 1);
+
+            Optional<EntityRecord> optional = storage.read(id, idMask);
+            assertTrue(optional.isPresent());
+            EntityRecord entityRecord = optional.get();
+
+            Message unpacked = unpack(entityRecord.getState());
+            assertFalse(isDefault(unpacked));
+        }
+
+        @SuppressWarnings("MethodWithMultipleLoops")
+        @Test
+        @DisplayName("multiple records according to the given field mask")
+        void multipleRecords() {
```

Comment:
Please consider refactoring this method into a nested class. It's way too complicated.